### PR TITLE
[GLUTEN-3302][VL] Fix dead lock of BackendJniWrapper static loading

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

BackendJniWrapper extend JniInitialized and JniInitializer invoke BackendJniWrapper#initializeBackend static method too, it will cause dead lock on static object lock.

PR #3929 change BackendJniWrapper#createExecutionCtx usage and let BackendJniWrapper init first in some case trigger this issue. Before 3929, JniInitializer's static lock was hold by other JniWrapper initialization which extend JniInitializer, it's safe to hold BackendJniWrapper static lock.

(Fixes: \#3302)

### CODE REVIEW

**Comment by Yohahaha** [comment]
@zhztheplayer

**Comment by zhztheplayer** [comment]
> @zhztheplayer How about this?

Very convincing, thank you very much for catching.

File: gluten-data/src/main/java/io/glutenproject/init/BackendJniWrapper.java [hunk 1]
```diff
@@ -16,13 +16,9 @@
  */
 package io.glutenproject.init;
 
-public class BackendJniWrapper extends JniInitialized {
+public class BackendJniWrapper {
```
Comment (zhztheplayer) [inline gluten-data/src/main/java/io/glutenproject/init/BackendJniWrapper.java:19]: I am feeling that we can remove this jni wrapper then move the method to class `JniInitialized`.

Then we can rename `JniInitialized` to `NativeBackendInitialized` or something to make things clearer.

How do you think about it?

Comment (Yohahaha) [reply gluten-data/src/main/java/io/glutenproject/init/BackendJniWrapper.java:19]: LGTM, let's do refactor in follow up patch.

I would like remove class hierarchy of JniInitialized too, expose a initBackend method and invoke this in ContextInitializer Object scala class, like below:
```scala
class ContextInitializer {
  override def initialize() = {
    init
    // ...
  }
}

object ContextInitializer {
  val init = JniInitialized.initBackend
}

```

Comment (zhztheplayer) [reply gluten-data/src/main/java/io/glutenproject/init/BackendJniWrapper.java:19]: > LGTM, let's do refactor in follow up patch.
> 
> I would like remove class hierarchy of JniInitialized too, expose a initBackend method and invoke this in ContextInitializer Object scala class, like below:
> 
> ```scala
> class ContextInitializer {
>   override def initialize() = {
>     init
>     // ...
>   }
> }
> 
> object ContextInitializer {
>   val init = JniInitialized.initBackend
> }
> ```
> 
> ContextInitializer's init variable will init in `<clinit>` as well as static initialization, thus only execute once.

Sounds good to me. Thanks.

**Review by zhztheplayer — APPROVED** [review]
Thanks!

**Comment by zhouyuan** [comment]
@zhztheplayer @Yohahaha  it looks like the TPCDS test in github action is green when merging the previous patch, is it due to the scale factor too small? can we add a gate check on this? 

thanks, -yuan

**Comment by Yohahaha** [comment]
> @zhztheplayer @Yohahaha it looks like the TPCDS test in github action is green when merging the previous patch, is it due to the scale factor too small? can we add a gate check on this?
> 
> thanks, -yuan

I found that issue in TPCDS sf1000 and can not repro every time, not sure how to add check.

**Comment by zhztheplayer** [comment]
It's a random issue related to JVM's class loading order which is naturally randomized. Class loading happens only once typically in one JVM process that why it's more reproduceable in cluster. UT will hardly cover the case unless we enable forked processes to run UTs.

I think adding a GHA local-cluster job could help either and we can try adding one at a good time.

File: gluten-data/src/main/java/io/glutenproject/init/BackendJniWrapper.java (hunk 1 above)
Comment (Yohahaha) [reply gluten-data/src/main/java/io/glutenproject/init/BackendJniWrapper.java:19]: #3307

**Pull request merged**